database/pgsql/pgsql.go

```diff
@@ -81,46 +86,161 @@ type Queryer interface {
 
 type pgSQL struct {
 	*sql.DB
-	cache *lru.ARCCache
+	cache  *lru.ARCCache
+	config Config
 }
 
+// Close closes the database and destroys if ManageDatabaseLifecycle has been specified in
+// the configuration.
 func (pgSQL *pgSQL) Close() {
-	pgSQL.DB.Close()
+	if pgSQL.DB != nil {
+		pgSQL.DB.Close()
+	}
+
+	if pgSQL.config.ManageDatabaseLifecycle {
+		dbName, pgSourceURL, _ := parseConnectionString(pgSQL.config.Source)
+		dropDatabase(pgSourceURL, dbName)
+	}
 }
 
+// Ping verifies that the database is accessible.
 func (pgSQL *pgSQL) Ping() bool {
 	return pgSQL.DB.Ping() == nil
 }
 
-// Open creates a Datastore backed by a PostgreSQL database.
-//
-// It will run immediately every necessary migration on the database.
-func Open(config *config.DatabaseConfig) (database.Datastore, error) {
-	// Run migrations.
-	if err := migrate(config.Source); err != nil {
-		log.Error(err)
-		return nil, database.ErrCantOpen
+// Config is the configuration that is used by openDatabase.
+type Config struct {
+	Source    string
+	CacheSize int
+
+	ManageDatabaseLifecycle bool
+	FixturePath             string
+}
+
+// openDatabase opens a PostgresSQL-backed Datastore using the given configuration.
+// It immediately every necessary migrations. If ManageDatabaseLifecycle is specified,
+// the database will be created first. If FixturePath is specified, every SQL queries that are
+// present insides will be executed.
+func openDatabase(registrableComponentConfig config.RegistrableComponentConfig) (database.Datastore, error) {
+	var pg pgSQL
+	var err error
+
+	// Parse configuration.
+	pg.config = Config{
+		CacheSize: 16384,
+	}
+	bytes, err := yaml.Marshal(registrableComponentConfig.Options)
+	if err != nil {
+		return nil, fmt.Errorf("pgsql: could not load configuration: %v", err)
+	}
+	err = yaml.Unmarshal(bytes, &pg.config)
+	if err != nil {
+		return nil, fmt.Errorf("pgsql: could not load configuration: %v", err)
+	}
+
+	dbName, pgSourceURL, err := parseConnectionString(pg.config.Source)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create database.
+	if pg.config.ManageDatabaseLifecycle {
+		log.Info("pgsql: creating database")
+		if err := createDatabase(pgSourceURL, dbName); err != nil {
+			return nil, err
+		}
 	}
 
 	// Open database.
-	db, err := sql.Open("postgres", config.Source)
+	pg.DB, err = sql.Open("postgres", pg.config.Source)
 	if err != nil {
-		log.Error(err)
-		return nil, database.ErrCantOpen
+		pg.Close()
+		return nil, fmt.Errorf("pgsql: could not open database: %v", err)
+	}
+
+	// Verify database state.
+	if err := pg.DB.Ping(); err != nil {
+		pg.Close()
```

Comment:
I meant the scenario where ping fails because the connection dropped.

It's irrelevant because pq returns an error when the connection is already closed, rather than panicing.